Sources/StreamChat/Utils/ThreadSafeWeakCollection.swift

```diff
@@ -0,0 +1,46 @@
+//
+// Copyright © 2022 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+class ThreadSafeWeakCollection<T: AnyObject> {
+    private let queue = DispatchQueue(label: "io.stream.com.weak-collection")
+    private let storage = NSHashTable<T>.weakObjects()
+
+    var allObjects: [T] {
+        var objects: [T]!
+        queue.sync {
+            objects = storage.allObjects
+        }
+        return objects
+    }
+
+    var count: Int {
+        var count: Int = 0
+        queue.sync {
+            count = storage.count
+        }
+        return count
+    }
+
+    func add(_ object: T?) {
```

Comment:
It's a bit unclear what we expect adding `nil` object to the collection. Can we make the param non optional?
```suggestion
    func add(_ object: T) {
```

Reply:
I am matching the signature of NSHashMap methods